### ISSUE TITLE
CSS Order Differs Between Development & Production Modes when Treeshaking.

### ISSUE THREAD

**Issue by brycehill** [webpack/webpack#7094]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
CSS seems to be out of order in "development" vs "production" mode. It appears somehow related to `sideEffects` and tree shaking. 

**If the current behavior is a bug, please provide the steps to reproduce.**

Here's a repo demonstrating the issue: https://github.com/brycehill/webpack-css-order.

There are instructions in the readme. Basically, if you build in `development` mode, the outputted CSS is different from `production` mode. 

**What is the expected behavior?**
CSS order is consistent from one environment to another. 

I don't want to depend on a specific CSS order, but I do expect it to be consistent from when I develop to when I build for production.

**If this is a feature request, what is motivation or use case for changing the behavior?**
N/A

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

Node: v8.9.1
Webpack: v4.6.0
OS: MacOS High Sierra

**Comment by brycehill** [comment]
Thanks for the reply. 

> using sideEffects you say order doesn't matter

I'm curious why this is the case. Is it a technical limitation? For me, the final order doesn't matter, but what does matter is a _consistent_ order between development and production _environments_

**Comment by TheLarkInn** [comment]
@brycehill are you saying that unused CSS which causes the development mode to look different from production mode is getting treeshaken?

**Comment by brycehill** [comment]
@TheLarkInn That's what I thought was happening at first. 

However, if you take a look at my example [repo](https://github.com/brycehill/webpack-css-order), none of the CSS is eliminated, **but** setting `sideEffects: true` fixes the problem. To reiterate from my previous comment:

> For me, the final order doesn't matter, but what _does_ matter is a consistent order between development and production environments

**Comment by sokra** [comment]
Should be fixed by https://github.com/webpack/webpack/pull/7507 in 4.12.1

**Comment by Fandekasp** [comment]
Not sure if this is related, but after upgrading babel to v7 (already on webpack 4.17.2), my staging/production webpack builds are missing any css including through `import "./style.less"` calls.
After some tedious debugging, I found out that adding the cli option `-p` or the config `mode: "production",` (or both) will break the build (which means that dev build works without any problem).

Since I had `"sideEffects": false,` in my packages.json, I changed it to `"sideEffects": ["*.css", "*.less"],`, and it seems to have fixed my build.

Can someone explain me what this sideEffects is doing, why it only affects production builids, and why it was working with Babel v6 and not with babel v7 ?

**Comment by cvle** [comment]
@sokra I just hit this issue in 4.27.1. It was also reproducable in https://github.com/brycehill/webpack-css-order with the upgraded webpack version.

Because of this I can't set `"sideEffects": false` nor `"sideEffects": ["*.css"]` in my `package.json`, otherwise it affects the CSS-order which causes styling issues.

**Comment by zamb3zi** [comment]
I'm also seeing this in 4.27.1.

**Comment by FuzzyTree** [comment]
I'm seeing this in 4.29. When I remove the 'sideEffects' property from on my vendor libraries the css order between production and dev are the same but when the property is included they are not.

**Comment by brycehill** [comment]
Looks like this was fixed in v4.12.1 as described by @sokra, however it broke again in the next minor release (v4.13.0). 

I pushed an update to my test [repo](https://github.com/brycehill/webpack-css-order) showing this.

**Comment by bz2** [comment]
@sokra Did this get deliberately reverted, or can the same fix just be reapplied to master?

**Comment by 7iomka** [comment]
v4.30.0 - broken css order when you import your css files in js
v4.31.0 - seems css order is as expected!

**Comment by 7iomka** [comment]
Attention, as before, when switching from one mode to another, for some strange reason, order gets lost!
To fix this you need to do `npm cache clear --force`
Why?

**Comment by Javey** [comment]
When import file from node_modules which has declared `"sideEffects": [".*css"]` in `package.json`, the css order is unexpected, but it's ok when import file from root directory.

I create a test repo: https://github.com/Javey/webpack-tree-shaking-demo

**Comment by darylteo** [comment]
@webpack-bot  please don't. This is an important issue and I'm disappointed we still haven't been able to solve without copious amounts of "!important" rules.

**Comment by evilebottnawi** [comment]
@darylteo don't worry we check all PRs what bot closed

**Comment by LoserAntbear** [comment]
It's pretty obvious, but might help someone.
TL;DR:
 For multiple entries configuration. This might be caused by scripts order in your `index.html`.

Hi guys. I'm not sure, that what i'm going to say is fully applicable to covered situation, but i had the same issue. 
In my React project i had 3 bundled `.js` files (`vendors.bundle.js`, `first.bundle.js`, `second.bundle.js`). We've been using style modules (imported styles).
When i was working in devenv, everything was cool. But after building for production styles become messy. Some was overwritten by another ones and so on.

The reason was obvious and i am a bit shy about not figuring out this in first minutes of investigation:
Scripts load order. Which was:
```
...
<script src="./vendors.bundle.js" type="text/javascript"></script>
<script src="./second.bundle.js" type="text/javascript"></script>
<script src="./first.bundle.js" type="text/javascript"></script>
...
```
All i had to do is switch order for last two guys.
Hope it might help someone like me. Cheers.

**Comment by Hypnosphi** [comment]
> using sideEffects you say order doesn't matter

Does it mean that I have to mark all JS files that import CSS as having side effects?

**Comment by artf** [comment]
Basically, in a real production application, with the use of code splitting for dynamic components (indispensable for us to keep the application bundle low) you just CAN'T use tree-shaking, which is a big loss.
 
Is there any chance to increase the priority of this issue? 🙏

**Comment by ebs-PriyankaKanwar** [comment]
Hello Team
I have facing issues regarding the next-CSS. I have using next-CSS but they are giving me a warning and when I removed the next-CSS from next.config file. my style CSS has not loaded chunks in the production mode. and my CSS is out of order in a production build.
Can you plz help me.

**Comment by ebs-PriyankaKanwar** [comment]
Do you want to request a feature or report a bug?
Bug

What is the current behavior?
CSS seems to be out of order in the "production" mode. And my style CSS has not loaded chunks in the production mode, and my CSS is out of order in a production build.

What is the expected behavior?
 My style CSS has not loaded chunks in the production mode.
here is my build
+ First Load JS shared by all                                  107 kB
  ├ static/pages/_app.js                                       16.1 kB
  ├ chunks/1899c3d08f1d4a9ab6e4e22050e66cd1600464e8.b9ea19.js  3.53 kB
  ├ chunks/48183758.ede116.js                                  65 B
  ├ chunks/57a5805a715b88beb8aec3ac43d8f962ef8879cc.53c4e4.js  2.34 kB
  ├ chunks/9ed67fcaced240be487e7a2880bfff868999eacb.76e83a.js  8.29 kB
  ├ chunks/ab68c4ca07cacf4e3048ee2d6f69810106b4b489.68a8b3.js  10.4 kB
  ├ chunks/d32b6132cd0996134c9f87f7274f553f791cd563.18dde5.js  6.39 kB
  ├ chunks/d6e4d54ae3ad4e326943499c7720eb830c61721b.b7693f.js  4.41 kB
  ├ chunks/framework.4dd100.js                                 40.3 kB
  ├ runtime/main.d3f0e8.js                                     13.2 kB
  ├ runtime/webpack.361a6a.js                                  1.59 kB
  ├ css/16c4518f007189356065.css                               54.9 kB
  └ css/f93669f8200e9712e425.css                               30.6 kB

Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.

Node: v13.7.0
Webpack:4.43.0
OS: Linux
Can you please help me.
Thanks

**Comment by Herz3h** [comment]
I'm not sure if it is related to this issue. But in certain parts of my code, this happens for JS files. When using `webpack-dev-server` it is out of order, but when I use `webpack` it is in order.

**Comment by otakustay** [comment]
Also having this issue in `5.0.0-beta.29`

**Comment by secondfry** [comment]
Well, if you need something fast and dirty you may apply [webpack+4.46.0.patch.txt](https://github.com/webpack/webpack/files/6359836/webpack%2B4.46.0.patch.txt) via `patch-package`...
The patch most probably breaks something somewhere since it makes `HarmonyImportSideEffectDependency` to behave exactly as `HarmonyImportDependency`.
But it makes development and production CSS oder to be consistent for me.

**Comment by vankop** [comment]
@alexander-akait is it still an issue for `webpack@5`?

**Comment by alexander-akait** [comment]
@vankop need tests, I think no

**Comment by thoughtspile** [comment]
In case you're curious, I have apparently worked around the issue with a Babel transform that traverses the dependencies and adds transitive CSS imports to every JS file in depth-first order: https://github.com/VKCOM/VKUI/pull/1933

Not sure if it's a stable fix though.

**Comment by marr** [comment]
> Well, if you need something fast and dirty you may apply [webpack+4.46.0.patch.txt](https://github.com/webpack/webpack/files/6359836/webpack%2B4.46.0.patch.txt) via `patch-package`... The patch most probably breaks something somewhere since it makes `HarmonyImportSideEffectDependency` to behave exactly as `HarmonyImportDependency` and removes `dependencyReference` hook call in `Compilation.js`. But it makes development and production CSS order to be consistent for me.

Note, that completely broke my build. Console error:
> TypeError: (void 0) is not a function

I was trying this as a workaround suggested here: https://github.com/vuetifyjs/vuetify/issues/13732

**Comment by marr** [comment]
Fwiw, I had been using `treeShake: false` in my nuxtjs-vuetify module settings due to a separate issue I was experiencing before. Removing that fixed the CSS order issue. To fix the other issue, I had to import a component I was rendering conditionally with `import { VCard } from 'vuetify/lib'`.

**Comment by william-will-angi** [comment]
I got around this issue by adding `sideEffects: true` to all of my webpack loader configs. Not the most ideal as it disable tree-shaking appwide. Hoping someone may have found a better solution.

**Comment by hlyst21** [comment]
@alexander-akait Greetings! Is there currently any way to solve the problem with the order of styles when using tree shaking?
I saw that the task has been moved to high priority in webpack 5/6, but if this can somehow be solved now, that would be just great.

My case:
On the react project.js using webpack 5.
Styling via css modules from under scss.
That is, each component has its own css module, for example:
```
component
  Component.js
  Component.module.scss
```  

```
// Component.js
import React from 'react'

import styles from './Component.module.scss'

const Component = () => {}
```

When you enable tree shaking in webpack (specify sideEffects: ["/.css", "/.scss"]), problems with css formation begin, namely, the order of modules entering the final css file ceases to be predictable and logical.
For example, there are two react components.js, Title and Content. In the Content component, we want to reuse the Title component, and pass other styles to it via className prop:
```
components
  title
    Title.js
    Title.module.scss
  content
    Content.js
    Content.module.scss
```

```
// Title.js
import React from 'react'
import cx from 'classnames'

import styles from './Title.module.scss'

const Title = ({ className }) => (
  <h1 className={cx(styles.title, className)}>Title</h1>
)

// Title.module.scss
.title {
  background: red;
}

// Content.js
import React from 'react'

import Title from '../title/Title'

import styles from './Content.module.scss'

const Content = ({ className }) => (
  <Title className={styles.title}></Title>
)

// Content.module.scss
.title {
  background: blue;
}
```

With tree shaking turned off, we will have the following final css file:
```
.components_title_title {background: red;} 
.components_content_title {background: blue;} 
```

That is, everything will work as expected, first come the styles of the Title component, then the styles of the Content component. The passed class from Content to Title will be below, so styles can be easily redefined.

However, when tree shaking is active, the style order will be the opposite of what we expect, i.e.:
```
.components_content_title {background: blue;}
.components_title_title {background: red;} 
```

In this case, the styles from the Content will not be able to override the styles of the Title, respectively.
This is very demotivating, because I want to take advantage of tree shaking, but because of this problem with styles, I can’t do it.

**Comment by alexander-akait** [comment]
Yeah, it is a bug inside webpack, I don't look deeply at this right now, so no solution right now, you can try to solve it on the CSS side, i.e. use css specificity, I know it is not a perfect solution, also you can try to use CSS modules

**Comment by kosciolek** [comment]
@alexander-akait I'd like to fix it myself if possible, would you accept a PR? I haven't done any contributions to webpack yet though. What do you think is the scope of this issue? Any suggestions as to where I should start investigating?

https://github.com/webpack/webpack/blob/6ccd5314451f392593a293f4dfd10153dc33c020/lib/buildChunkGraph.js#L792 It looks like webpack traverses modules here?

**Comment by alexander-akait** [comment]
@kosciolek I'm afraid this is not an easy task, you can start with adding failed test cases and then we will look how to fix it

**Comment by kosciolek** [comment]
One of the commits in 5.90.2 seems to have fixed it, at least for [the reproduction I made above](https://github.com/kosciolek/webpack-css-order-reproduction). However, our larger project still has this problem, but it's likely it's something on our side.

**Comment by alexander-akait** [comment]
@kosciolek Yeah, we can also have some edge cases, so, if you can provide an example I will investigate it

> One of the commits in 5.90.2 seems to have fixed it

Release will be today/tomorrow, need to finish one small thing

**Comment by kosciolek** [comment]
This is the commit in 5.90.2 that fixed the issue in the reproduction:
```
commit a2ce375d968a9ef5e8b9ee59a30268d3b35c864a
Merge: 2a063f88d 1673ee5d1
Author: Alexander Akait <4567934+alexander-akait@users.noreply.github.com>
Date:   Mon Feb 12 13:37:05 2024 +0300

    fix: use new runtime to reconsider skipped connections activeState
```

**Comment by kosciolek** [comment]
I made a [PR with a failed test case](https://github.com/webpack/webpack/pull/18302)

**Comment by dudorovdmitrii** [comment]
Any update on this? I have this problem as well. The uikit I'm using has {"sideEffects": ["*.css", "*.scss"],} in its package.json. 

Code:
```
import { Button } from "@gravity-ui/uikit";
import "./main.css";

ReactDOM.createRoot(document.getElementById("root")).render(
    <Button>sdf</Button>
);
```

Css build result:
```
// Css from main.css
.main {
  color: green;
}

// Button css
.g-button {
  --_--text-color: var(--g-color-text-primary);
}
```

**Comment by receter** [comment]
One solution to avoid this problem as a component library author is to add all styles in a component library package to a CSS layer. In this way, the styles in the consuming application always take precedence over the styles in the library package no matter what the order of appearance is.

But keep in mind that style rules with less specificity will have precedence as well. So if you have some sort of CSS reset in addition to the library package, you might want to add this to a CSS layer as well and define the order early in the application.

**Comment by alexander-akait** [comment]
Close in favor - https://github.com/webpack/webpack/issues/18961
